### DISCUSSION TITLE
AutoDisposeProviderBase<Future<bool>> does not work with ref.listen()

### DISCUSSION THREAD

**Discussion by abegehr** [rrousselGit/riverpod#1075] [Q&A]
**Describe the bug**
The FutureProvider<bool> isAuthenticatedProvider resolves to a boolean with represents if the user is authenticated or not. This provider is listened to on the WelcomeScreen to navigate to the Home Screen once the user is authenticated.

**To Reproduce**

This is the isAuthenticatedProvider provider:

```
final isAuthenticatedProvider = FutureProvider.autoDispose<bool>((ref) async {
  final user = await ref.watch(userProvider.future);
  final isAuthenticated = user != null;
  debugPrint('isAuthenticatedProvider – yield $isAuthenticated');
  return isAuthenticated;
});
```

This call to ref.listen from WelcomeScreen.build() does _not_ work:

```
ref.listen<Future<bool>>(
      isAuthenticatedProvider.future,
      (_, value) async {
        if (await value) {
          debugPrint(" === authenticated.");
          Navigator.popUntil(context, ModalRoute.withName('/'));
        }
      },
      onError: (error, stackTrace) {
        debugPrint(" === ERROR: $error");
      },
    );
```

This call to ref.listen from WelcomeScreen.build() does work:

```
ref.listen<AsyncValue<bool>>(
      isAuthenticatedProvider,
      (_, value) {
        value.whenData((isAuthenticated) {
          if (isAuthenticated) {
            debugPrint(" === authenticated.");
            Navigator.popUntil(context, ModalRoute.withName('/'));
          }
        });
      },
      onError: (error, stackTrace) {
        debugPrint(" === ERROR: $error");
      },
    );
```

Neither print any error to the console.

**Expected behavior**
I expect `ref.listen(isAuthenticatedProvider.future, …)` to work.

**Comment by rrousselGit** [comment]
What does "does not work" mean exactly? What's the problem?

**Comment by abegehr** [comment]
The callback function does not get called. I also Troyes adding a debug print before "if (await value)" but that does not print to the console either.

**Comment by nbourdin** [comment]
you can try with `ref.listen<bool?>(isAuthenticatedProvider.select((e) => e.value, …)`

**Comment by rrousselGit** [comment]
Try adding a "fireImmediately: true" to `ref.listen`

Also consider using `when` instead of `whenData`. It's possible that your provider is in loading or error state
